# NumPy Function Summary with Examples

| Function  | Description                                                                                     | Example Code                                     |
|-----------|-------------------------------------------------------------------------------------------------|--------------------------------------------------|
| `random`  | Generates random numbers in a given shape.                                                      | `np.random.random((2,2))`                         |
| `rand`    | Produces an array of random samples from a uniform distribution over `[0, 1)`.                  | `np.random.rand(3, 2)`                            |
| `randn`   | Returns samples from the "standard normal" distribution with mean 0 and variance 1.             | `np.random.randn(3, 3)`                           |
| `randint` | Returns random integers from a specified low to high range.                                     | `np.random.randint(1, 10, size=(2, 3))`           |
| `arange`  | Creates an array with evenly spaced values within a given interval.                             | `np.arange(0, 10, 2)`                             |
| `arange`  | Creates an array with a range                                                                   | `np.arange(5)`                                    |
| `ones`    | Generates an array of the specified shape filled with ones.                                     | `np.ones((2, 3))`                                 |
| `zeros`   | Similar to `ones`, but fills the array with zeros.                                              | `np.zeros((2, 3))`                                |
| `eye`     | Constructs a 2-D array with ones on the diagonal and zeros elsewhere (identity matrix).         | `np.eye(3)`                                       |
| `shape`   | Returns the dimensions of an array as a tuple.                                                  | `arr.shape` where `arr = np.array([[1, 2], [3, 4]])` |
| `reshape` | Reshapes an array without changing its data.                                                    | `arr.reshape((4, 1))` where `arr = np.array([1, 2, 3, 4])` |
| `linspace`| Generates an array of evenly spaced numbers over a specified interval.                          | `np.linspace(0, 1, 5)`                            |

# NumPy Function Summary with Outputs

| Function  | Example Code                                     | Output                               |
|-----------|--------------------------------------------------|--------------------------------------|
| `random`  | `np.random.random((2,2))`                        | `array([[0.68, 0.22], [0.33, 0.73]])` (values will vary) |
| `rand`    | `np.random.rand(3, 2)`                           | `array([[0.12, 0.85], [0.44, 0.68], [0.53, 0.95]])` (values will vary) |
| `randn`   | `np.random.randn(3, 3)`                          | `array([[-0.38, 0.91, -1.05], [0.39, -0.32, 0.75], [-0.15, -0.67, 0.47]])` (values will vary) |
| `randint` | `np.random.randint(1, 10, size=(2, 3))`          | `array([[5, 2, 9], [4, 8, 6]])` (values will vary) |
| `arange`  | `np.arange(0, 10, 2)`                            | `array([0, 2, 4, 6, 8])`             |
| `arange`  | `np.arange(5)`                            | `array([0, 1, 2, 3, 4])`             |
| `ones`    | `np.ones((2, 3))`                                | `array([[1., 1., 1.], [1., 1., 1.]])`|
| `zeros`   | `np.zeros((2, 3))`                               | `array([[0., 0., 0.], [0., 0., 0.]])`|
| `eye`     | `np.eye(3)`                                      | `array([[1., 0., 0.], [0., 1., 0.], [0., 0., 1.]])` |
| `shape`   | `arr.shape` where `arr = np.array([[1, 2], [3, 4]])` | `(2, 2)`                           |
| `reshape` | `arr.reshape((4, 1))` where `arr = np.array([1, 2, 3, 4])` | `array([[1], [2], [3], [4]])` |
| `linspace`| `np.linspace(0, 1, 5)`                           | `array([0., 0.25, 0.5, 0.75, 1.])`   |

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
    <style>
        h3 {
            font-family: monospace;
            color: #DCD427
        }
        li {
            font-family: monospace;
            color: #DCD427;
            font-size: 20px
        }
        strong {
            font-size: 16px;
            color: #c55955
        }
    </style>
</head>
<body>
    <h3>Course topic</h3>
    <div>
        <li>arrange <strong>(one dimentional array)</strong></li>
        <li>zeros <strong>(one/two dimentional array)</strong></li>
        <li>ones <strong>(one/two dimentional array)</strong></li>
        <li>linspace <strong>(one dimentional array)</strong></li>
        <li>eye <strong>(one/two dimentional array)</strong></li>
    </div>
</body>
</html>

In [2]:
import numpy as np

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
    <style>
        h3 {
            font-family: monospace;
            color: #DCD427;
        }
    </style>
</head>
<body>
    <h3>arrange, zeros, ones</h3>
</body>
</html>

In [3]:
np.arange(0,10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [4]:
np.arange(20)

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19])

In [5]:
np.arange(0,11,2)

array([ 0,  2,  4,  6,  8, 10])

In [6]:
np.zeros(3)

array([0., 0., 0.])

In [7]:
np.zeros((5,3))

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

In [8]:
np.ones(3)

array([1., 1., 1.])

In [9]:
np.ones((3,4))

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

In [10]:
np.ones((3,4)) + np.ones((3,4))

array([[2., 2., 2., 2.],
       [2., 2., 2., 2.],
       [2., 2., 2., 2.]])

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
    <style>
        h3 {
            font-family: monospace;
            color: #DCD427;
        }
    </style>
</head>
<body>
    <h3>linspace</h3>
</body>
</html>

<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Math Calculation</title>
  <style>
    p {
      font-family: monospace;
      padding: 20px;
      background-color: black;
      color: white;
      font-size: 16px;
      color: #DCD427;
    }
    h4 {
      color: #00FFFF;
    }
  </style>
</head>
<body>
  <h4>Detailed Breakdown:</h4>
  <div>
    <p> 
        <b> Calculate the Step Size:</b>
        <br>
        The step size can be calculated by taking the difference between the stop and start values, and then dividing this by num - 1 (which is the number of intervals).
        In your case, <strong>(1 - 0) / (10 - 1) = 1 / 9 ≈ 0.11111111</strong>.
    </p>
  </div>
</body>
</html>



In [11]:
# Generate 10 numbers between 0 and 1, not including 1
linear_spaced_numbers_without_endpoint = np.linspace(0, 1, num=10)
print(linear_spaced_numbers_without_endpoint)

[0.         0.11111111 0.22222222 0.33333333 0.44444444 0.55555556
 0.66666667 0.77777778 0.88888889 1.        ]


<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Step Breakdown</title>
  <style>
    body {
      font-family: monospace;
      padding: 20px;
      background-color: black;
      color: white;
      font-size: 5px;
    }
    .step {
      font-family: monospace;
      margin-bottom: 10px;
      padding: 10px;
      background-color: #222;
      border-left: 4px solid #00BFFF;
    }
    .step strong {
      display: inline-block;
      width: 110px;
      color: #DCD427;
    }
    h4 {
      font-family: monospace;
      color: #00FFFF;
    }
  </style>
</head>
<body>
  <h4>Detailed Breakdown:</h4>
  <div class="step"><strong>First value:</strong> 0 (start point)</div>
  <div class="step"><strong>Second value:</strong> 0 + 1 × 1⁄9 ≈ 0.11111111</div>
  <div class="step"><strong>Third value:</strong> 0 + 2 × 1⁄9 ≈ 0.22222222</div>
  <div class="step"><strong>Fourth value:</strong> 0 + 3 × 1⁄9 ≈ 0.33333333</div>
  <div class="step"><strong>Fifth value:</strong> 0 + 4 × 1⁄9 ≈ 0.44444444</div>
  <div class="step"><strong>Sixth value:</strong> 0 + 5 × 1⁄9 ≈ 0.55555555</div>
  <div class="step"><strong>Seventh value:</strong> 0 + 6 × 1⁄9 ≈ 0.66666666</div>
  <div class="step"><strong>Eighth value:</strong> 0 + 7 × 1⁄9 ≈ 0.77777777</div>
  <div class="step"><strong>Ninth value:</strong> 0 + 8 × 1⁄9 ≈ 0.88888888</div>
  <div class="step"><strong>Tenth value:</strong> 0 + 9 × 1⁄9 = 1.0 (end point)</div>
</body>
</html>


In [12]:
linear_spaced_numbers = np.linspace(0, 5, 100)
print(linear_spaced_numbers)

[0.         0.05050505 0.1010101  0.15151515 0.2020202  0.25252525
 0.3030303  0.35353535 0.4040404  0.45454545 0.50505051 0.55555556
 0.60606061 0.65656566 0.70707071 0.75757576 0.80808081 0.85858586
 0.90909091 0.95959596 1.01010101 1.06060606 1.11111111 1.16161616
 1.21212121 1.26262626 1.31313131 1.36363636 1.41414141 1.46464646
 1.51515152 1.56565657 1.61616162 1.66666667 1.71717172 1.76767677
 1.81818182 1.86868687 1.91919192 1.96969697 2.02020202 2.07070707
 2.12121212 2.17171717 2.22222222 2.27272727 2.32323232 2.37373737
 2.42424242 2.47474747 2.52525253 2.57575758 2.62626263 2.67676768
 2.72727273 2.77777778 2.82828283 2.87878788 2.92929293 2.97979798
 3.03030303 3.08080808 3.13131313 3.18181818 3.23232323 3.28282828
 3.33333333 3.38383838 3.43434343 3.48484848 3.53535354 3.58585859
 3.63636364 3.68686869 3.73737374 3.78787879 3.83838384 3.88888889
 3.93939394 3.98989899 4.04040404 4.09090909 4.14141414 4.19191919
 4.24242424 4.29292929 4.34343434 4.39393939 4.44444444 4.4949

In [13]:
# Generate numbers between 0 and 1, returning the step size as well
linear_spaced_numbers_with_step = np.linspace(0, 1, num=5, retstep=True)
print(linear_spaced_numbers_with_step)

(array([0.  , 0.25, 0.5 , 0.75, 1.  ]), np.float64(0.25))


In [14]:
linear_spaced_numbers_with_step = np.linspace(0, 5, num=10)
print(linear_spaced_numbers_with_step)

[0.         0.55555556 1.11111111 1.66666667 2.22222222 2.77777778
 3.33333333 3.88888889 4.44444444 5.        ]


<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
    <style>
        h3 {
            font-family: monospace;
            color: #DCD427;
        }
    </style>
</head>
<body>
    <h3>eye</h3>
</body>
</html>

In [15]:
# Diagonal matrix with 1s on the diagonal
np.eye(5)

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])